From f6b5105d65b25842e3ca6c6df4c2f392e6a3a5d9 Mon Sep 17 00:00:00 2001
From: Brian Smith <brian@dbsoft.org>
Date: Wed, 31 Jan 2024 09:36:41 -0600
Subject: [PATCH 02/14] Issue #2051 - Follow-up Part 1b: Fix building on later
 MacOS versions. Clang complains about duplicate definition of
 GfxMatrixToCGAffineTransform(). CGFontGetGlyphPath() is non-public and
 removed in later MacOS versions. CTFontCreatePathForGlyph() is the CoreText
 replacement in 10.5 and later.

---
 gfx/2d/DrawTargetSkia.cpp | 13 -------------
 gfx/2d/ScaledFontMac.cpp  |  6 ++++++
 2 files changed, 6 insertions(+), 13 deletions(-)

diff --git a/gfx/2d/DrawTargetSkia.cpp b/gfx/2d/DrawTargetSkia.cpp
index 49732a6ddd..e1e44bf630 100644
--- a/gfx/2d/DrawTargetSkia.cpp
+++ b/gfx/2d/DrawTargetSkia.cpp
@@ -968,19 +968,6 @@ private:
   CGContextRef mCG;
 };
 
-static inline CGAffineTransform
-GfxMatrixToCGAffineTransform(const Matrix &m)
-{
-  CGAffineTransform t;
-  t.a = m._11;
-  t.b = m._12;
-  t.c = m._21;
-  t.d = m._22;
-  t.tx = m._31;
-  t.ty = m._32;
-  return t;
-}
-
 /***
  * We have to do a lot of work to draw glyphs with CG because
  * CG assumes that the origin of rects are in the bottom left
diff --git a/gfx/2d/ScaledFontMac.cpp b/gfx/2d/ScaledFontMac.cpp
index c478c1226c..a9aec4fcba 100644
--- a/gfx/2d/ScaledFontMac.cpp
+++ b/gfx/2d/ScaledFontMac.cpp
@@ -94,7 +94,13 @@ ScaledFontMac::GetPathForGlyphs(const GlyphBuffer &aBuffer, const DrawTarget *aT
           // XXX: we could probably fold both of these transforms together to avoid extra work
           CGAffineTransform flip = CGAffineTransformMakeScale(1, -1);
 
+          // CGFontGetGlyphPath is non-public and CoreText can do the same on 10.5 and later
+#if !defined(MAC_OS_X_VERSION_10_5) || (MAC_OS_X_VERSION_MAX_ALLOWED < MAC_OS_X_VERSION_10_5)
           CGPathRef glyphPath = ::CGFontGetGlyphPath(mFont, &flip, 0, aBuffer.mGlyphs[i].mIndex);
+#else
+          CTFontRef ctFont = CTFontCreateWithGraphicsFont(mFont, 0, NULL, NULL);
+          CGPathRef glyphPath = CTFontCreatePathForGlyph(ctFont, aBuffer.mGlyphs[i].mIndex, &flip);
+#endif
 
           CGAffineTransform matrix = CGAffineTransformMake(mSize, 0, 0, mSize,
                                                            aBuffer.mGlyphs[i].mPosition.x,
-- 
2.50.1

